File: rubyapp1/game.rb
# frozen_string_literal: true

# 推箱子核心逻辑（Ruby 教学）

class GameState
  attr_reader :walls, :goals, :boxes, :player, :moves, :won, :width, :height, :level_index

  def self.from_rows(rows, index = 0)
    new(rows, index)
  end

  def initialize(rows, index = 0)
    @walls = {}
    @goals = {}
    @boxes = {}
    @player = [0, 0]
    @moves = 0
    @won = false
    @level_index = index
    @hist = []
    max_x = 0
    max_y = 0
    rows.each_with_index do |row, y|
      max_y = y
      row.each_char.with_index do |ch, x|
        max_x = x if x > max_x
        k = key(x, y)
        case ch
        when '#' then @walls[k] = true
        when '.' then @goals[k] = true
        when '$' then @boxes[k] = true
        when '*'
          @boxes[k] = true
          @goals[k] = true
        when '@' then @player = [x, y]
        when '+'
          @player = [x, y]
          @goals[k] = true
        end
      end
    end
    @width = max_x + 1
    @height = max_y + 1
  end

  def try_move(dx, dy)
    return false if @won

    px, py = @player
    nx = px + dx
    ny = py + dy
    nk = key(nx, ny)
    return false if @walls[nk]

    if @boxes[nk]
      bx = nx + dx
      by = ny + dy
      bk = key(bx, by)
      return false if @walls[bk] || @boxes[bk]

      @hist << { player: @player.dup, box_from: nk, box_to: bk }
      @boxes.delete(nk)
      @boxes[bk] = true
      @player = [nx, ny]
      @moves += 1
      check_win
      return true
    end

    @hist << { player: @player.dup, box_from: nil, box_to: nil }
    @player = [nx, ny]
    true
  end

  def undo
    return false if @won || @hist.empty?

    entry = nil
    until @hist.empty?
      entry = @hist.pop
      break if entry[:box_from]

      @player = entry[:player]
    end
    return true if entry.nil? || entry[:box_from].nil?

    @player = entry[:player]
    @boxes.delete(entry[:box_to])
    @boxes[entry[:box_from]] = true
    @moves -= 1 if @moves.positive?
    @won = false
    true
  end

  def render_ascii
    lines = []
    @height.times do |y|
      row = +''
      @width.times do |x|
        k = key(x, y)
        row << if @player == [x, y]
                 @goals[k] ? '+' : '@'
               elsif @boxes[k]
                 @goals[k] ? '*' : '$'
               elsif @walls[k]
                 '#'
               elsif @goals[k]
                 '.'
               else
                 ' '
               end
      end
      lines << row
    end
    "#{lines.join("\n")}\n"
  end

  private

  def key(x, y)
    "#{x},#{y}"
  end

  def check_win
    @won = @boxes.keys.all? { |b| @goals[b] }
  end
end

File: juliaapp1/game.jl
# 推箱子核心逻辑（Julia 教学）

const Cell = String

mutable struct Hist
    player::Tuple{Int,Int}
    box_from::Union{Nothing,Cell}
    box_to::Union{Nothing,Cell}
end

mutable struct GameState
    walls::Set{Cell}
    goals::Set{Cell}
    boxes::Set{Cell}
    player::Tuple{Int,Int}
    moves::Int
    won::Bool
    width::Int
    height::Int
    hist::Vector{Hist}
end

cell_key(x::Int, y::Int) = string(x, ",", y)

function from_rows(rows::Vector{String}, index::Int=0)::GameState
    walls = Set{Cell}()
    goals = Set{Cell}()
    boxes = Set{Cell}()
    px, py = 0, 0
    max_x, max_y = 0, 0
    for (y, row) in enumerate(rows)
        max_y = y - 1
        for (x0, ch) in enumerate(row)
            x = x0 - 1
            max_x = max(max_x, x)
            k = cell_key(x, y - 1)
            if ch == '#'
                push!(walls, k)
            elseif ch == '.'
                push!(goals, k)
            elseif ch == '$'
                push!(boxes, k)
            elseif ch == '*'
                push!(boxes, k); push!(goals, k)
            elseif ch == '@'
                px, py = x, y - 1
            elseif ch == '+'
                px, py = x, y - 1
                push!(goals, k)
            end
        end
    end
    GameState(walls, goals, boxes, (px, py), 0, false, max_x + 1, max_y + 1, Hist[])
end

function check_win!(s::GameState)
    s.won = all(b -> b in s.goals, s.boxes)
end

function try_move!(s::GameState, dx::Int, dy::Int)::Bool
    s.won && return false
    px, py = s.player
    nx, ny = px + dx, py + dy
    nk = cell_key(nx, ny)
    nk in s.walls && return false
    if nk in s.boxes
        bx, by = nx + dx, ny + dy
        bk = cell_key(bx, by)
        (bk in s.walls || bk in s.boxes) && return false
        push!(s.hist, Hist(s.player, nk, bk))
        delete!(s.boxes, nk)
        push!(s.boxes, bk)
        s.player = (nx, ny)
        s.moves += 1
        check_win!(s)
        return true
    end
    push!(s.hist, Hist(s.player, nothing, nothing))
    s.player = (nx, ny)
    true
end

function undo!(s::GameState)::Bool
    (s.won || isempty(s.hist)) && return false
    while !isempty(s.hist)
        e = pop!(s.hist)
        if e.box_from !== nothing
            s.player = e.player
            delete!(s.boxes, e.box_to)
            push!(s.boxes, e.box_from)
            s.moves > 0 && (s.moves -= 1)
            s.won = false
            return true
        end
        s.player = e.player
    end
    true
end

function render_ascii(s::GameState)::String
    io = IOBuffer()
    for y in 0:s.height-1
        for x in 0:s.width-1
            k = cell_key(x, y)
            if s.player == (x, y)
                print(io, k in s.goals ? '+' : '@')
            elseif k in s.boxes
                print(io, k in s.goals ? '*' : '$')
            elseif k in s.walls
                print(io, '#')
            elseif k in s.goals
                print(io, '.')
            else
                print(io, ' ')
            end
        end
        print(io, '\n')
    end
    String(take!(io))
end
